### ISSUE TITLE
A bug in solid voxelization

### ISSUE THREAD

**Issue by AaronZhangY** [Forceflow/cuda_voxelizer#64]
I try to input 32/50/64/100 as voxel grid length in cpu mode.
32/64/100 work very well, but 50 has a bug.
I have uploaded my source obj file and output .obj file (modify .txt to .obj).
<img width="494" alt="1642051507(1)" src="https://user-images.githubusercontent.com/80893143/149270879-5bbc1393-0e15-4b59-802b-405d996a2efb.png">
[sphere.txt](https://github.com/Forceflow/cuda_voxelizer/files/7860152/sphere.txt)
[sphere.obj_50_voxels.txt](https://github.com/Forceflow/cuda_voxelizer/files/7860153/sphere.obj_50_voxels.txt)

**Comment by Forceflow** [comment]
If I read this correctly, you're trying a non-cubical voxel grid? There is still no official support for this - ymmv.

Thanks for report, though.

**Comment by AaronZhangY** [comment]
I will try to developer branch, thx.
And I notice in the paper Fast Parallel Surface and Solid Voxelization on GPU's, they use octree in solid voxelization.
Do you have use octree in this project?

**Comment by AaronZhangY** [comment]
Another bug happened in solid voxelization.
I use grid size = 8 in cpu mode.
My input is a cube, and output is 8x8x7. I think the correct output is 8x8x8.
Do you have any idea to solve this problem?
![1643107787(1)](https://user-images.githubusercontent.com/80893143/150963502-51cbe67b-2822-449c-b0af-8a3c789d50a9.png)

**Comment by xingshh** [comment]
I find this bug too。 I think the boundary pnt should be reserved., then reset the bounary bit?
when  x = xmax; 
{
							if (morton_order) {
								size_t location = mortonEncode_LUT(x, y, z);
								setBitXor((voxel_table, location);
								boundaryBits.push_back(location);
							}else {
								size_t location = static_cast<size_t>(x) + (static_cast<size_t>(y) * static_cast<size_t>(info.gridsize.x))+  (static_cast<size_t>(z) * static_cast<size_t>(info.gridsize.x) * static_cast<size_t>(info.gridsize.y);
setBitXor(voxel_table, location);
								boundaryBits.push_back(location);
							}
						} 
}

**Comment by conceptclear** [comment]
您好，您的邮件我已收到并会尽快回复，祝好

**Comment by Forceflow** [comment]
Can you try again with the latest release version?

**Comment by hitrpr** [comment]
Tested now .5 release, have the same single-voxel layer.
127 width for main object and 127 grid with 2 layers for additional
![изображение](https://user-images.githubusercontent.com/21099918/192902603-b12f0254-8dc9-4c91-af99-c89130665972.png)
I did  -s 128

**Comment by Forceflow** [comment]
And this only happens in solid voxelisation?

**Comment by adnan-saood** [comment]
I can do a basic workaround to produce what I want but in a very unintuitive way, If I use an STL input file with mm units (scaled x1000 compared to the zip file I attached earlier), the result is good. I do not know what caused this, maybe some numerical errors.

![image](https://user-images.githubusercontent.com/22562330/227257596-b8724680-7b63-4f2f-be69-0305dca7be3f.png)

**Comment by Forceflow** [comment]
Yeah, I'm writing this down as some numerical errors. I'm currently in the process of rewriting the project to remove the dependency on GLM, maybe this resolves these numerical quirks, since it is not really meant for usage with CUDA.

**Comment by adnan-saood** [comment]
@Forceflow I will definitely try that. If this is a collaboration-based library don't hesitate to create the task and maybe we would do the changes in our forks and PR. Thanks again for the clarifications.

**Comment by Forceflow** [comment]
You can test with the dev branch if you want.
